// @nxrocks/common/src/lib/testing/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.syncToAsyncIterable = exports.mockZipEntries = exports.mockExecutorContext = void 0;
const tslib_1 = require("tslib");
function mockExecutorContext(pluginName, executorName, workspaceVersion = 2) {
    return {
        root: '/root',
        cwd: 'root',
        projectName: 'proj',
        workspace: {
            npmScope: 'nxrocks',
            version: workspaceVersion,
            projects: {
                proj: {
                    root: 'proj',
                    targets: {
                        test: {
                            executor: `${pluginName}:${executorName}`,
                        },
                    },
                },
            },
        },
        target: {
            executor: `${pluginName}:${executorName}`,
        },
        isVerbose: false,
    };
}
exports.mockExecutorContext = mockExecutorContext;
function mockZipEntries(files) {
    return files.map((e) => {
        return {
            path: typeof e === 'string' ? e : e.filePath,
            buffer: () => { var _a; return Promise.resolve(Buffer.from(typeof e === 'string' ? e : (_a = e.fileContent) !== null && _a !== void 0 ? _a : e.filePath)); },
            type: 'File',
        };
    });
}
exports.mockZipEntries = mockZipEntries;
function syncToAsyncIterable(syncIterable) {
    return tslib_1.__asyncGenerator(this, arguments, function* syncToAsyncIterable_1() {
        for (const elem of syncIterable) {
            yield yield tslib_1.__await(elem);
        }
    });
}
exports.syncToAsyncIterable = syncToAsyncIterable;
